### ISSUE TITLE
Compatibility with Numpy breaks across minor versions of Python

### ISSUE THREAD

**Issue by Dominik1123** [matplotlib/matplotlib#16696]
### Bug report

**Bug summary**

I'm using `matplotlib==2.1.1` with `numpy==1.15.0` on Python 3.6 and everything works fine. However when installing the same package versions on Python 3.7 I get an import error related to numpy. I checked that numpy actually has the module that is reported to be missing (and why should it be missing since it's the same version), so the problem must be with matplotlib. The error occurs when doing `import matplotlib.backends.backend_qt5agg`.

I'm wondering why this issue occurs when upgrading my version of Python but using the same package versions. I see that both matplotlib and numpy have dedicated compiled wheels for each Python version, but I'd expect the behavior to remain similar since the package version doesn't change.

Is this expected / intended? Can I expect similar behavior for other versions of matplotlib as well?

**Code for reproduction**

```
$ conda create -n test-py37 python=3.7
$ conda activate test-py37
$ pip install matplotlib==2.1.1 numpy==1.15.0
$ python
Python 3.7.6 (default, Jan  8 2020, 19:59:22) 
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import matplotlib.backends.backend_qt5agg
```

**Actual outcome**

```
ModuleNotFoundError: No module named 'numpy.core._multiarray_umath'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/dominik/miniconda3/envs/test-py37/lib/python3.7/site-packages/matplotlib/backends/backend_qt5agg.py", line 13, in <module>
    from matplotlib.transforms import Bbox
  File "/home/dominik/miniconda3/envs/test-py37/lib/python3.7/site-packages/matplotlib/transforms.py", line 42, in <module>
    from matplotlib._path import (affine_transform, count_bboxes_overlapping_bbox,
ImportError: numpy.core.multiarray failed to import
```

**Expected outcome**

This is the behavior on Python 3.6. It can be reproduced by following the above steps and specifying `python=3.6` when creating the environment. The error message here occurs because PyQt is not installed and it will go away once it is installed. The point is that here no incomtability with Numpy is reported.

```
Traceback (most recent call last):
  File "/home/dominik/miniconda3/envs/test-py36/lib/python3.6/site-packages/matplotlib/backends/qt_compat.py", line 198, in <module>
    from PySide import QtCore, QtGui, __version__, __version_info__
ModuleNotFoundError: No module named 'PySide'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/dominik/miniconda3/envs/test-py36/lib/python3.6/site-packages/matplotlib/backends/backend_qt5agg.py", line 16, in <module>
    from .backend_qt5 import (
  File "/home/dominik/miniconda3/envs/test-py36/lib/python3.6/site-packages/matplotlib/backends/backend_qt5.py", line 18, in <module>
    import matplotlib.backends.qt_editor.figureoptions as figureoptions
  File "/home/dominik/miniconda3/envs/test-py36/lib/python3.6/site-packages/matplotlib/backends/qt_editor/figureoptions.py", line 20, in <module>
    import matplotlib.backends.qt_editor.formlayout as formlayout
  File "/home/dominik/miniconda3/envs/test-py36/lib/python3.6/site-packages/matplotlib/backends/qt_editor/formlayout.py", line 56, in <module>
    from matplotlib.backends.qt_compat import QtGui, QtWidgets, QtCore
  File "/home/dominik/miniconda3/envs/test-py36/lib/python3.6/site-packages/matplotlib/backends/qt_compat.py", line 201, in <module>
    "Matplotlib qt-based backends require an external PyQt4, PyQt5,\n"
ImportError: Matplotlib qt-based backends require an external PyQt4, PyQt5,
PySide or PySide2 package to be installed, but it was not found.
```

**Matplotlib version**

```
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
ca-certificates           2020.1.1                      0  
certifi                   2019.11.28               py37_0  
cycler                    0.10.0                   pypi_0    pypi
ld_impl_linux-64          2.33.1               h53a641e_7  
libedit                   3.1.20181209         hc058e9b_0  
libffi                    3.2.1                hd88cf55_4  
libgcc-ng                 9.1.0                hdf63c60_0  
libstdcxx-ng              9.1.0                hdf63c60_0  
matplotlib                2.1.1                    pypi_0    pypi
ncurses                   6.2                  he6710b0_0  
numpy                     1.15.0                   pypi_0    pypi
openssl                   1.1.1d               h7b6447c_4  
pip                       20.0.2                   py37_1  
pyparsing                 2.4.6                    pypi_0    pypi
python                    3.7.6                h0371630_2  
python-dateutil           2.8.1                    pypi_0    pypi
pytz                      2019.3                   pypi_0    pypi
readline                  7.0                  h7b6447c_5  
setuptools                45.2.0                   py37_0  
six                       1.14.0                   pypi_0    pypi
sqlite                    3.31.1               h7b6447c_0  
tk                        8.6.8                hbc83047_0  
wheel                     0.34.2                   py37_0  
xz                        5.2.4                h14c3975_4  
zlib                      1.2.11               h7b6447c_3
```

  * Operating system: `Ubuntu 16.04 LTS (4.4.0-173-generic #203-Ubuntu SMP Wed Jan 15 02:55:01 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux)`
  * Matplotlib version: `2.1.1`
  * Matplotlib backend (`print(matplotlib.get_backend())`): `TkAgg`
  * Python version: `3.7.6`
  * Jupyter version (if applicable): `--`
  * Other libraries: `numpy==1.15.0`

`pip install matplotlib==2.1.1 numpy==1.15.0`.

**Comment by QuLogic** [comment]
Why do you need to use such an old version of Matplotlib with Python 3.7? Python 3.7 did not exist when it was released, so there are no wheels of 2.1.1. The Matplotlib you get is compiled from source, and when `pip` builds it, it uses an isolated copy of NumPy, not the one you asked for. You can get it to work by installing NumPy first in a separate step (but also clear out any cached wheels that you may have built.)

**Comment by Dominik1123** [comment]
I froze the dependencies of one of my projects since I have no time to maintain it anymore. These were the two most recent versions of numpy and matplotlib by that time. I didn't think that upgrading to newer versions of Python would break the compatibility (unless Python introduces breaking changes itself). When a user reported the error I became aware of it.

Indeed Python 3.7 was released after Matplotlib 2.1.1 but I though when installing via `pip` it will sort out the dependencies and use the version of Numpy that I specified. Here is the output I get:

```
$ pip install matplotlib==2.1.1 numpy==1.15.0 --no-cache-dir
Collecting matplotlib==2.1.1
  Downloading matplotlib-2.1.1.tar.gz (36.1 MB)
     |████████████████████████████████| 36.1 MB 1.4 MB/s 
Collecting numpy==1.15.0
  Downloading numpy-1.15.0-cp37-cp37m-manylinux1_x86_64.whl (13.8 MB)
     |████████████████████████████████| 13.8 MB 6.5 MB/s 
Collecting six>=1.10
  Downloading six-1.14.0-py2.py3-none-any.whl (10 kB)
Collecting python-dateutil>=2.0
  Downloading python_dateutil-2.8.1-py2.py3-none-any.whl (227 kB)
     |████████████████████████████████| 227 kB 8.9 MB/s 
Collecting pytz
  Downloading pytz-2019.3-py2.py3-none-any.whl (509 kB)
     |████████████████████████████████| 509 kB 8.1 MB/s 
Collecting cycler>=0.10
  Downloading cycler-0.10.0-py2.py3-none-any.whl (6.5 kB)
Collecting pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1
  Downloading pyparsing-2.4.6-py2.py3-none-any.whl (67 kB)
     |████████████████████████████████| 67 kB 10.6 MB/s 
Building wheels for collected packages: matplotlib
  Building wheel for matplotlib (setup.py) ... done
  Created wheel for matplotlib: filename=matplotlib-2.1.1-cp37-cp37m-linux_x86_64.whl size=10623672 sha256=872404474357a363bc1549b3d5795f6562526264aea519ae982bcee8d76d81b5
  Stored in directory: /tmp/pip-ephem-wheel-cache-1sswuypn/wheels/f3/7a/d6/021782cff10d8257e030d4a766ca5ed9667fd8758606fbbeff
Successfully built matplotlib
Installing collected packages: numpy, six, python-dateutil, pytz, cycler, pyparsing, matplotlib
Successfully installed cycler-0.10.0 matplotlib-2.1.1 numpy-1.15.0 pyparsing-2.4.6 python-dateutil-2.8.1 pytz-2019.3 six-1.14.0
```

First it downloads all the dependencies (where it downloads `numpy-1.15.0`) and then it builds `matplotlib` from source. This suggests that `numpy-1.15.0` is used and already satisfied as a requirement. There is no sign of any other version of Numpy being downloaded. So I don't understand what you mean when saying that `pip` uses an isolated copy of Numpy. Where would that copy come from?

**Comment by tacaswell** [comment]
`  Building wheel for matplotlib (setup.py) ... done`  <- that line is like likely hiding the logs where the other version of numpy is being installed.   I suspect what is happening is that you are installing 1.15 into your env, the latest version into the env where you build matplotlib, and then they disagree when the wheel from Matplotlib is installed into your env.  See https://pip.pypa.io/en/stable/reference/pip/#pep-517-and-518-support for more details about build isolation.

I think you can pass `--no-build-isolation` to pip.

I am going to close this and suggest that the discussion move to https://discourse.matplotlib.org/c/community/install/13 because I do not think there is anything we can do on the code side.
